plugins/builtin/strategy/threshold/plugin/plugin.go

```diff
@@ -0,0 +1,296 @@
+	defaultWithinBoundsTrigger = 5
+)
+
+var (
+	PluginID = plugins.PluginID{
+		Name:       pluginName,
+		PluginType: sdk.PluginTypeStrategy,
+	}
+
+	PluginConfig = &plugins.InternalPluginConfig{
+		Factory: func(l hclog.Logger) interface{} { return NewThresholdPlugin(l) },
+	}
+
+	pluginInfo = &base.PluginInfo{
+		Name:       pluginName,
+		PluginType: sdk.PluginTypeStrategy,
+	}
+)
+
+// thresholdPluginRunConfig are the parsed values for a threshold plugin run.
+type thresholdPluginRunConfig struct {
+	upperBound          float64
+	lowerBound          float64
+	actionType          string
+	actionValue         float64
+	withinboundsTrigger int
+}
+
+// Assert that StrategyPlugin meets the strategy.Strategy interface.
+var _ strategy.Strategy = (*StrategyPlugin)(nil)
+
+// StrategyPlugin is the Threshold implementation of the strategy.Strategy
+// interface.
+type StrategyPlugin struct {
+	logger hclog.Logger
+}
+
+// NewTargetValuePlugin returns the Threshold implementation of the
+// strategy.Strategy interface.
+func NewThresholdPlugin(log hclog.Logger) strategy.Strategy {
+	return &StrategyPlugin{
+		logger: log,
+	}
+}
+
+// SetConfig satisfies the SetConfig function on the base.Base interface.
+func (s *StrategyPlugin) SetConfig(_ map[string]string) error {
+	return nil
+}
+
+// PluginInfo satisfies the PluginInfo function on the base.Base interface.
+func (s *StrategyPlugin) PluginInfo() (*base.PluginInfo, error) {
+	return pluginInfo, nil
+}
+
+// Run satisfies the Run function on the strategy.Strategy interface.
+func (s *StrategyPlugin) Run(eval *sdk.ScalingCheckEvaluation, count int64) (*sdk.ScalingCheckEvaluation, error) {
+	if len(eval.Metrics) == 0 {
+		s.logger.Trace("no metrics available")
+		return nil, nil
+	}
+
+	// Parse check config.
+	config, err := parseConfig(eval.Check.Strategy.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := s.logger.With("check_name", eval.Check.Name, "current_count", count,
+		"lower_bound", config.lowerBound, "upper_bound", config.upperBound,
+		"actionType", config.actionType)
+
+	// Check if we have enough data points within bounds.
+	if !withinBounds(logger, eval.Metrics, config) {
+		logger.Trace("not enough data points within bounds")
+		return nil, nil
+	}
+
+	// Calculate new count.
+	logger.Trace("calculating new count")
+
+	var newCount int64
+	switch config.actionType {
+	case runConfigKeyDelta:
+		newCount = runDelta(count, config.actionValue)
+	case runConfigKeyPercentage:
+		newCount = runPercentage(count, config.actionValue)
+	case runConfigKeyValue:
+		newCount = runValue(count, config.actionValue)
+	}
+
+	logger = logger.With("new_count", newCount)
+
+	// Identify the direction of scaling, and exit early if none.
+	eval.Action.Direction = calculateDirection(count, newCount)
+	if eval.Action.Direction == sdk.ScaleDirectionNone {
+		return eval, nil
+	}
+
+	logger = logger.With("direction", eval.Action.Direction)
+	logger.Trace("calculated scaling strategy results")
```

Comment:
Two sub-loggers get created to log a single trace line here. I wonder if it would be better to add `new_count` and `direction` onto the log line as inline contextual arguments like `logger.Trace("msg", "new_count", newCount)`?

Reply:
I think there used to be a log message between the two, but I guess I got rid of it at some point 😅